bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ReadAheadManager.java

```diff
@@ -0,0 +1,655 @@
+                conf.getLong(READ_AHEAD_TASK_EXPIRED_TIME_MS, DEFAULT_READ_AHEAD_TASK_EXPIRED_TIME_MS),
+                conf.getLong(READ_AHEAD_TIMEOUT_MS, DEFAULT_READ_AHEAD_TIMEOUT_MS));
+    }
+
+    public ReadAheadManager(EntryLogger entryLogger, EntryLocationIndex entryLocationIndex,
+                            ReadCache cache, DbLedgerStorageStats dbLedgerStorageStats,
+                            boolean submitReadAheadTaskImmediately, int readAheadTaskPoolSize,
+                            int readAheadMessages, int readAheadBytes, double preTriggerReadAheadRatio,
+                            long readAheadTaskExpiredTimeMs, long readAheadTimeoutMs) {
+        // core components initialization
+        readAheadExecutor = Executors.newFixedThreadPool(
+                readAheadTaskPoolSize, new DefaultThreadFactory("read-ahead"));
+        cleanupExecutor = Executors.newSingleThreadScheduledExecutor(
+                new DefaultThreadFactory("read-ahead-cleanup"));
+
+        pendingReadAheadPositions = new ConcurrentHashMap<>();
+        pendingDeletePositions = new ConcurrentLinkedQueue<>();
+
+        inProgressReadAheadTaskStatuses = new ConcurrentHashMap<>();
+        pendingDeleteReadAheadTaskStatuses = new ConcurrentLinkedQueue<>();
+
+        // external assistant components assignment
+        this.entryLogger = entryLogger;
+        this.entryLocationIndex = entryLocationIndex;
+        this.cache = cache;
+
+        // metrics
+        this.dbLedgerStorageStats = dbLedgerStorageStats;
+
+        // configurable arguments
+        this.submitReadAheadTaskImmediately = submitReadAheadTaskImmediately;
+        this.readAheadMessages = readAheadMessages;
+        this.readAheadBytes = readAheadBytes;
+        this.preTriggerReadAheadRatio = preTriggerReadAheadRatio;
+
+        this.readAheadTaskExpiredTimeMs = readAheadTaskExpiredTimeMs;
+        this.readAheadTimeoutMs = readAheadTimeoutMs;
+
+        cleanupExecutor.scheduleAtFixedRate(
+                SafeRunnable.safeRun(this::removeExpiredReadAheadTasks), 30, 30, TimeUnit.SECONDS);
+    }
+
+    public void shutdown() {
+        this.readAheadExecutor.shutdown();
+        this.cleanupExecutor.shutdown();
+    }
+
+    private static void recordStatsInNano(OpStatsLogger logger, long startTimeNanos) {
+        logger.registerSuccessfulEvent(MathUtils.elapsedNanos(startTimeNanos), TimeUnit.NANOSECONDS);
+    }
+
+    private static void recordStatsInNano(Counter counter, long startTimeNanos) {
+        counter.add(MathUtils.elapsedNanos(startTimeNanos));
+    }
+
+    public void addNextReadPosition(long expectedLedgerId, long expectedEntryId,
+                                    long actualStartLedgerId, long actualStartEntryId, long location) {
+        LedgerEntryPosition lep = new LedgerEntryPosition(expectedLedgerId, expectedEntryId);
+        pendingReadAheadPositions.put(lep, new ReadAheadPos(
```

Comment:
When does this map remove items?

Reply:
A separate cleanup thread will periodically removes expired read-ahead positions.